### PR TITLE
Remove resource state attributes that are no longer in the schema

### LINKED ISSUE #25752: Provider Removed Attributes Causing "data could not be decoded from the state: unsupported attribute" Error
### Terraform Version

```
v0.13.0-rc1
```

Although its also being reported with v0.12.29.

### Terraform Configuration Files

main.tf

```terraform
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "2.70.0"
    }
  }

  required_version = "0.13.0"
}

provider "aws" {
  region = "us-east-2"
}

module "test" {
  source = "./testmodule"
}

output "region" {
  value = module.test.region_name
}
```

testmodule/main.tf:

```terraform
data "aws_region" "current" {}

output "region_name" {
  value = data.aws_region.current.name
}

```

### Debug Output

Please ask if you need it.

### Expected Behavior

Since there are no references to any removed attributes, no errors after upgrading the provider.

### Actual Behavior

```
Error: Invalid resource instance data in state

  on testmodule/main.tf line 1:
   1: data "aws_region" "current" {}

Instance module.test.data.aws_region.current data could not be decoded from
the state: unsupported attribute "current".
```

### Steps to Reproduce

1. `terraform init`
1. `terraform apply`
1. Change `version = "2.70.0"` to `version = "3.0.0"`
1. `terraform init`
1. `terraform apply`

### References

* https://github.com/terraform-providers/terraform-provider-aws/issues/13398
* https://github.com/terraform-providers/terraform-provider-aws/pull/14001
* https://github.com/terraform-providers/terraform-provider-aws/issues/14431
* https://github.com/terraform-providers/terraform-provider-aws/issues/14458

### PR DESCRIPTION
**Pull request by jbardin**
While removal of attributes can be handled by providers through the
UpgradeResourceState call, data sources may need to be evaluated before
reading, and they have no upgrade path in the provider protocol.
Strip out extra attributes during state decoding when they are no longer
present in the schema, and there is no schema upgrade pending. If there 
is a pending schema upgrade, let the provider handle it since it may want to
check for the removed attribute values. 

Another option not explored here is declaring that providers must migrate
any attributes pending removal before they actually removed, which would
allow the possibility of making the decoder less strict and always strip out
unknown attributes (similar to `encoding/json`). Since we can
handle the case fully within core, it seems reasonable to let the decoder be
as strict as possible, though it might be useful to implement the functionality
upstream for other projects using zclconf.

The validate command should work only with the configuration, but when
validate was run at the start of a plan or apply command the state was
inserted in preparation for the next walk. This could lead to errors
when the resource schemas had changes and the state could not be
upgraded or decoded.

Fixes #25752

### CODE REVIEW

File: command/testdata/apply-vars/main.tf [hunk 1]
```diff
@@ -1,5 +1,5 @@
 variable "foo" {}
 
 resource "test_instance" "foo" {
-    value = "${var.foo}"
+    value = var.foo
```
Comment (pselle) [inline command/testdata/apply-vars/main.tf:4]: Nice update

**Comment by reverendtimm** [comment]
Is there any ETA on .13.1 release? Gasping for my TF stacks to work again 👍

**Comment by renan** [comment]
Yes, please release a new version!

**Comment by ChristophShyper** [comment]
So now, if I import some resource that was created in other way and possibly modified, I have no clue which attributes were set to which values? If I destroy it and recreate it with Terraform then this resource can have different attributes' values.

That is not a bugfix, that is a breaking change and should be loudly advertised.

**Pull request merged**